src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,229 @@
+  circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w) :=
+begin
+  ext i j,
+  simp only [mul_apply, mul_vec, circulant, dot_product],
+  refine fintype.sum_equiv (equiv.sub_right j) _ _ _,
+  intro x,
+  simp only [equiv.sub_right_apply],
+  congr' 2,
+  abel
+end
+
+lemma fin.circulant_mul [semiring α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant (mul_vec (circulant v) w)
+| 0     := dec_trivial
+| (n+1) := circulant_mul
+
+/-- circulant matrices commute in multiplication under certain condations. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype I] [add_comm_group I] (v w : I → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub I] [has_scalar R α] {k : R} {v : I → α} :
+  circulant (k • v) = k • circulant v :=
+by {ext, simp [circulant]}
+
+@[simp] lemma circulant_zero [has_zero α] [has_sub I] :
+  circulant 0 = (0 : matrix I I α) :=
+by ext; simp [circulant]
+
+@[simp] lemma circulant_single [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  circulant (pi.single 0 1 : I → α) = (1 : matrix I I α) :=
+by { ext i j, simp only [circulant, one_apply, pi.single_apply, sub_eq_zero], congr }
+
+/-- Any identity matrix is a circulant matrix. -/
+lemma one_eq_circulant [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, ite (i = 0) 1 0) :=
+by { rw [←circulant_single], congr, ext i, simp [pi.single, update] }
+
+/-- Another version of `one_eq_circulant`. -/
+lemma one_eq_circulant' [has_zero α] [has_one α] [decidable_eq I] [add_group I] :
+  (1 : matrix I I α) = circulant (λ i, (1 : matrix I I α) i 0) :=
+by { ext i j, simp only [circulant, one_apply, sub_eq_zero], congr }
+
+lemma fin.one_eq_circulant [has_zero α] [has_one α] :
+  ∀ n, (1 : matrix (fin n) (fin n) α) = circulant (λ i, ite (↑i = 0) 1 0)
+| 0     := dec_trivial
+| (n+1) :=
+begin
+  ext i j,
+  simp only [circulant, one_apply, ← @fin.coe_zero n, fin.coe_injective.eq_iff, sub_eq_zero],
+  congr,
+end
+
+/-- For a one-ary predicate `p`, `p` applied to every entry of `circulant v` is true
+    if `p` applied to every entry of `v` is true. -/
+lemma pred_circulant_entry_of_pred_vec_entry [has_sub I]
+  {p : α → Prop} {v : I → α} (h : ∀ k, p (v k)) (i j : I) :
+  p ((circulant v) i j) := h (i - j)
```

Comment:
This lemma seems pointless to me; surely anyone faced with this kind of goal will just do `rw circulant` and it becomes trivial?

Reply:
I can remove it

Reply:
Yes, I think you should